test/gpu/native/studies/sort/radixSortGpu.chpl

```diff
@@ -0,0 +1,288 @@
+            prefixSum[i+2*offset-1] = prefixSum[i+2*offset-1] + t;
+          }
+          offset = offset >> 1;
+        }
+        hostprefixSum = prefixSum;
+      }
+  }
+
+  proc arrScan(ref count: [?D] uint, ref inputArr: [] uint, exp: int, bitMask: int, buckets: int){
+    // Count the number of occurrences of each bitsAtATime-bit chunk in the array
+    // at the current exp position
+    on here.gpus[useGpuId]{
+      var gpuCount : [D] uint; // Initialized all 0s
+      var gpuInputArr = inputArr;
+      @assertOnGpu
+      foreach i in gpuInputArr {
+        gpuAtomicAdd(gpuCount[((i>>exp) & bitMask):int], 1);
+        // count[((i>>exp) & bitMask):int] += 1;
+      }
+      count = gpuCount;
+    }
+    if scanTest then
+      writeln(" Count: ", count);
+
+
+    if scanType == "hillisSteele" then
+      hillisSteeleScan(count, buckets);
+    else if scanType == "blelloch" then
+      blellochScan(count, buckets);
+    else if scanType == "serial" then
+      serialScan(count, buckets);
+
+    if scanTest then
+      writeln("Pf sum: ",count);
+  }
+
+  proc radixSort(inputArr: [] uint){
+    if (noisy){
+        writeln("Bits at a time: ", bitsAtATime);
+        writeln("Buckets: ", buckets);
+        writeln("Bit mask: ", bitMask);
+        writeln("Input: ", inputArr);
+    }
+
+    // Create the output array and copy input
+    var outputArr: [0..<inputArr.size] uint;
+
+    // Create the count and prefix sum arrays once
+    // This was we can reuse it for each iteration of radix Sort
+    var prefixSum: [0..<buckets] uint;
+
+    // Ceiling on number of iterations based on max element in array
+    // Reduce can be used to do this faster
+    // But I'm leaving this as is due to https://github.com/chapel-lang/chapel/issues/22736
+    var max = inputArr[0];
+    for i in 1..<inputArr.size {
+      if inputArr[i] > max {
+        max = inputArr[i];
+      }
+    }
+
+    // Number of iterations is the number of bits in the max element divided by number of bits we sort at a time
+    var exp = 0;
+    while (max> 0){
+      if(noisy) then
+        writeln("       Exp: ", exp);
+
+      arrScan(prefixSum, inputArr, exp, bitMask, buckets);
+
+      if noisy then
+        writeln("       Prefix sum: ", prefixSum);
+      // Use the prefix sum to reorder the array at the exp position
+      // This is done by iterating through the array backwards
+      // and placing each element in the correct position in the output array
+
+
+      // on here.gpus[useGpuId]{
+      //   var gpuPrefixSum = prefixSum;
+      //   var gpuInputArr = inputArr;
+      //   var gpuOutputArr : outputArr.type;
+      //   var gpuExp = exp;
+      //   @assertOnGpu
+      //   foreach i in gpuInputArr{
+      //     const idx = ((i>>gpuExp) & bitMask):int;
+      //     const arrIdx = gpuAtomicAdd(gpuPrefixSum[idx], 1);
+      //     gpuOutputArr[arrIdx:int] = i;
+      //   }
+      //   outputArr = gpuOutputArr;
+      // }
+
```

Comment:
The commented out code is racey and doesn't work.
Why is the above code racey?
Let's work through an example
Say we have 10 elements in the array
And we sort 3 bits at a time (i.e. 8 buckets)
Say the input array looks like this:
[70, 75, 12, 80, 20, 24, 90, 43, 45, 504]
And we're sorting the last 3 bits. Therefore the prefix sum looks like this:
[0, 5, 9, 9, 9, 9, 9, 9] // The algorithm calculated this. It's correct, trust me.

So what happens?
The first element is 70. So we take the last 3 bits: 1.
The second element is 75. So we take the last 3 bits: 1.
Last 3 bits for all the elements is:
[1, 1, 0, 1, 0, 0, 1, 0, 0, 7]
So the first element goes in the 1st bucket. 
The second element goes in the 1st bucket
...
The last element goes in the 7th bucket.

At this point I've realized why the code is racey.
Ex:
We want the first element (70) to get arrIdx=5. In the serial version it will,
because it will access the prefix sum at idx=1 and get 5.
Next, the second element (75) will then get arrIdx=6.
      
But here, the second element can get there first and get arrIdx=5 since it's a foreach loop.
And then the first element will get arrIdx=6. This means the array won't be sorted.

So actually the best we can do is to use a foreach loop in chunks where we don't need the serialism.
In this example we can divide up the array's each unique idx (0, 1, 7 in this case) needs to be run in serial.
But the three can each be run individually in parallel because that's not  going to be racey across different idx's.